File: implementations/go/provekit-lift-go/rpc.go
package liftgo

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"os"
)

type rpcRequest struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      json.RawMessage `json:"id"`
	Method  string          `json:"method"`
	Params  json.RawMessage `json:"params"`
}

type liftParams struct {
	WorkspaceRoot string         `json:"workspace_root"`
	SourcePaths   []string       `json:"source_paths"`
	Options       map[string]any `json:"options"`
}

// dialectOptionsFromParams reads the verify-facing dialect selector from the
// lift `options` bag. The verifier-driving caller (kit-dispatch) passes
// `options.dialect = "core"` so the body-derived postcondition uses SMT-LIB
// core op symbols and is z3-dischargeable. Absent / any other value keeps the
// round-trip namespaced dialect.
func dialectOptionsFromParams(opts map[string]any) LiftOptions {
	if opts == nil {
		return LiftOptions{}
	}
	if d, ok := opts["dialect"].(string); ok && d == "core" {
		return LiftOptions{NormalizeCoreArith: true}
	}
	return LiftOptions{}
}

func RunRPC(stdin io.Reader, stdout io.Writer) error {
	return RunRPCWithDefault(stdin, stdout, LiftOptions{})
}

// RunRPCWithDefault runs the lift RPC loop with a default op dialect applied
// to every `lift` call that does not itself name a `dialect` in its options.
// The kit-dispatch verify pipeline resolves the `go` lift surface to a binary
// launched with `--dialect=core`, so the body-derived postconditions are
// z3-dischargeable without the language-neutral dispatcher having to know
// anything about Go.
func RunRPCWithDefault(stdin io.Reader, stdout io.Writer, defaultOpts LiftOptions) error {
	scanner := bufio.NewScanner(stdin)
	for scanner.Scan() {
		line := scanner.Bytes()
		if len(line) == 0 {
			continue
		}
		var req rpcRequest
		if err := json.Unmarshal(line, &req); err != nil {
			writeRPC(stdout, errorResponse(nil, -32700, "PARSE_ERROR"))
			continue
		}
		switch req.Method {
		case "initialize":
			writeRPC(stdout, successResponse(req.ID, InitializeResult()))
		case "lift":
			writeRPC(stdout, handleLift(req.ID, req.Params, defaultOpts))
		case "compile":
			writeRPC(stdout, handleCompile(req.ID, req.Params))
		case "shutdown":
			writeRPC(stdout, successResponse(req.ID, nil))
			return nil
		default:
			writeRPC(stdout, errorResponse(req.ID, -32601, fmt.Sprintf("METHOD_NOT_FOUND: %s", req.Method)))
		}
	}
	return scanner.Err()
}

func handleLift(id json.RawMessage, raw json.RawMessage, defaultOpts LiftOptions) any {
	var params liftParams
	if len(raw) > 0 {
		if err := json.Unmarshal(raw, &params); err != nil {
			return errorResponse(id, -32602, "invalid lift params")
		}
	}
	if params.WorkspaceRoot == "" {
		cwd, err := os.Getwd()
		if err == nil {
			params.WorkspaceRoot = cwd
		} else {
			params.WorkspaceRoot = "."
		}
	}
	if len(params.SourcePaths) == 0 {
		return errorResponse(id, -32602, "source_paths must be a non-empty array of strings")
	}
	liftOpts := defaultOpts
	if perCall := dialectOptionsFromParams(params.Options); perCall.NormalizeCoreArith {
		liftOpts = perCall
	}
	result, err := LiftPathsWithOptions(params.WorkspaceRoot, params.SourcePaths, liftOpts)
	if err != nil {
		return errorResponse(id, -32603, fmt.Sprintf("Lift failed: %v", err))
	}
	return successResponse(id, map[string]any{
		"kind":          "ir-document",
		"ir":            result.IR,
		"callEdges":     []any{},
		"diagnostics":   result.Diagnostics,
		"opacityReport": []any{},
		"refusals":      result.Refusals,
	})
}

func handleCompile(id json.RawMessage, raw json.RawMessage) any {
	var params struct {
		IR json.RawMessage `json:"ir"`
	}
	if err := json.Unmarshal(raw, &params); err != nil {
		return errorResponse(id, -32602, "invalid compile params")
	}
	irItems, err := sourceUnitIRFromJSON(params.IR)
	if err != nil {
		return errorResponse(id, -32602, "compile params.ir must be an array")
	}
	compiled, err := Compile(CompileInput{IR: irItems})
	if err != nil {
		return errorResponse(id, -32603, fmt.Sprintf("Compile failed: %v", err))
	}
	return successResponse(id, compiled)
}

func successResponse(id json.RawMessage, result any) map[string]any {
	return map[string]any{"jsonrpc": "2.0", "id": idJSONValue(id), "result": result}
}

func errorResponse(id json.RawMessage, code int, message string) map[string]any {
	return map[string]any{"jsonrpc": "2.0", "id": idJSONValue(id), "error": map[string]any{"code": code, "message": message}}
}

func idJSONValue(id json.RawMessage) any {
	if len(id) == 0 {
		return nil
	}
	dec := json.NewDecoder(bytes.NewReader(id))
	dec.UseNumber()
	var out any
	if err := dec.Decode(&out); err != nil {
		return nil
	}
	return out
}

func writeRPC(w io.Writer, v any) {
	b, err := marshalJSONNoHTML(v)
	if err != nil {
		fmt.Fprintf(w, `{"jsonrpc":"2.0","id":null,"error":{"code":-32603,"message":"%s"}}`+"\n", err.Error())
		return
	}
	fmt.Fprintln(w, string(b))
}

File: implementations/php/provekit-lift-php-source/src/Rpc.php
<?php

declare(strict_types=1);

namespace ProvekIt\LiftPhpSource;

function initialize_result(): array
{
    return [
        'name' => 'provekit-lift-php-source',
        'version' => VERSION,
        'protocol_version' => 'pep/1.7.0',
        'dialect' => DIALECT,
        'capabilities' => [
            'authoring_surfaces' => [DIALECT],
            'ir_version' => 'v1.1.0',
            'emits_signed_mementos' => false,
        ],
    ];
}

function run_rpc(): void
{
    $stdin = fopen('php://stdin', 'r');
    while (($line = fgets($stdin)) !== false) {
        $line = trim($line);
        if ($line === '') {
            continue;
        }
        try {
            $request = json_decode($line, true, 512, JSON_THROW_ON_ERROR);
            if (!is_array($request)) {
                write_rpc(error_response(null, -32600, 'INVALID_REQUEST'));
                continue;
            }
            write_rpc(dispatch_rpc($request));
        } catch (\JsonException $e) {
            write_rpc(error_response(null, -32700, 'PARSE_ERROR: ' . $e->getMessage()));
        } catch (\Throwable $e) {
            write_rpc(error_response($request['id'] ?? null, -32603, $e->getMessage()));
        }
    }
}

function dispatch_rpc(array $request): array
{
    $id = $request['id'] ?? null;
    $method = $request['method'] ?? '';
    $params = is_array($request['params'] ?? null) ? $request['params'] : [];
    return match ($method) {
        'initialize' => success_response($id, initialize_result()),
        'lift' => lift_rpc($id, $params),
        'compile' => compile_rpc($id, $params),
        'provekit.plugin.recognize' => recognize_rpc($id, $params),
        'shutdown' => success_response($id, null),
        default => error_response($id, -32601, 'METHOD_NOT_FOUND: ' . $method),
    };
}

function lift_rpc(mixed $id, array $params): array
{
    $surface = is_string($params['surface'] ?? null) ? $params['surface'] : DIALECT;
    if ($surface !== DIALECT) {
        return error_response($id, 1003, 'SURFACE_NOT_SUPPORTED: ' . $surface);
    }
    $sourcePaths = array_values(array_filter(
        is_array($params['source_paths'] ?? null) ? $params['source_paths'] : [],
        static fn(mixed $path): bool => is_string($path)
    ));
    if ($sourcePaths === []) {
        return error_response($id, -32602, 'source_paths must be a non-empty array of strings');
    }
    $workspaceRoot = is_string($params['workspace_root'] ?? null) ? $params['workspace_root'] : '.';
    $result = (new PhpSourceLifter())->liftPaths($workspaceRoot, $sourcePaths);
    return success_response($id, [
        'kind' => 'ir-document',
        'ir' => $result['ir'],
        'callEdges' => $result['callEdges'],
        'diagnostics' => $result['diagnostics'],
        'opacityReport' => $result['opacityReport'],
        'refusals' => $result['refusals'],
    ]);
}

function compile_rpc(mixed $id, array $params): array
{
    $ir = $params['ir'] ?? null;
    if (!is_array($ir)) {
        return error_response($id, -32602, 'ir must be an array of function-contract mementos');
    }
    return success_response($id, ['kind' => 'compiled-formula', 'body' => (new PhpSourceCompiler())->compileIrDocument($ir)]);
}

function recognize_rpc(mixed $id, array $params): array
{
    $projectRoot = $params['project_root'] ?? null;
    if (!is_string($projectRoot) || $projectRoot === '') {
        return error_response($id, -32602, 'project_root must be a non-empty string');
    }
    $sourcePaths = array_values(array_filter(
        is_array($params['source_paths'] ?? null) ? $params['source_paths'] : [],
        static fn(mixed $path): bool => is_string($path)
    ));
    $bindingTemplates = array_values(array_filter(
        is_array($params['binding_templates'] ?? null) ? $params['binding_templates'] : [],
        static fn(mixed $binding): bool => is_array($binding)
    ));

    try {
        return success_response($id, (new PhpSourceRecognizer())->recognizePaths($projectRoot, $sourcePaths, $bindingTemplates));
    } catch (\InvalidArgumentException $e) {
        return error_response($id, -32602, $e->getMessage());
    }
}

function success_response(mixed $id, mixed $result): array
{
    return ['jsonrpc' => '2.0', 'id' => $id, 'result' => $result];
}

function error_response(mixed $id, int $code, string $message): array
{
    return ['jsonrpc' => '2.0', 'id' => $id, 'error' => ['code' => $code, 'message' => $message]];
}

function write_rpc(array $response): void
{
    fwrite(STDOUT, json_encode($response, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE) . "\n");
}
